In [3]:
import pandas as pd
import numpy as np 
from scipy.stats import skew

In [5]:
print(''' dans cette etape je vais charger mes données que j'ai deja scrapper 
    et créer une data frame qui contient les données 
''')

file_path = "appartements_data.csv"
data_frame = pd.read_csv(file_path)

 dans cette etape je vais charger mes données que j'ai deja scrapper 
    et créer une data frame qui contient les données 



In [7]:
data_frame.head()

titre          prix  \
0  A louer Appartement prés clinique Badr(Bourgogne)      5 500 DH   
1            Appartement à vendre 132 m² a bv 2 mars  1 400 000 DH   
2  appartement 3 chambre salon résidence green parck      6 500 DH   
3  Apt H-Standing 160M²/Terrasse/Très Sécurisé/Ra...     11 000 DH   
4                     Appartement 112 m à mahaj ryad      8 000 DH   

  emplacement           type_de_vente  salon  chambre  toilette surface  \
0  Casablanca                 A louer    1.0      NaN       NaN     100   
1  Casablanca  Appartements, à vendre    1.0      3.0       2.0     132   
2        Salé                 A louer    1.0      NaN       NaN     116   
3  Casablanca                 A louer    1.0      NaN       NaN     160   
4       Rabat                 A louer    1.0      NaN       NaN     112   

                                                lien  
0  https://www.avito.ma/fr/bourgogne/appartements...  
1  https://www.avito.ma/fr/2_mars/appartements/Ap...  
2  https://www.avito.ma/fr/sala_el_jadida/apparte...  
3  https://www.avito.ma/fr/anfa/appartements/Apt_...  
4  https://www.avito.ma/fr/hay_riad/appartements/...

In [9]:
data_frame.describe()
data_frame['surface'] = pd.to_numeric(data_frame['surface'], errors='coerce')

In [11]:
# Convertir toutes les valeurs en chaînes
data_frame['prix'] = data_frame['prix'].astype(str)

# Remplacer les valeurs non spécifiées par NaN
data_frame['prix'] = data_frame['prix'].replace('Prix Non Specifié', np.nan)

# Remplacer les espaces insécables par des espaces normaux
data_frame['prix'] = data_frame['prix'].str.replace('\u202f', ' ', regex=True)

# Extraire uniquement la partie numérique en tenant compte des espaces et des symboles
data_frame['prix'] = data_frame['prix'].str.extract(r'([0-9\s]+)')[0]  # Capturer les chiffres, même avec des espaces

# Nettoyer les espaces et convertir en float
data_frame['prix'] = data_frame['prix'].str.replace(' ', '').astype(float)

In [13]:
data_frame.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 431 entries, 0 to 430
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   titre          431 non-null    object 
 1   prix           391 non-null    float64
 2   emplacement    431 non-null    object 
 3   type_de_vente  431 non-null    object 
 4   salon          272 non-null    float64
 5   chambre        134 non-null    float64
 6   toilette       131 non-null    float64
 7   surface        277 non-null    float64
 8   lien           431 non-null    object 
dtypes: float64(5), object(4)
memory usage: 30.4+ KB


In [15]:
# Documenting each column in the dataset as part of the data catalog
data_catalog = pd.DataFrame({
    "Column": data_frame.columns,
    "Description": [
        "Title of the property listing",
        "Price of the property (may include currency)",
        "Location of the property",
        "Type of transaction (e.g., for rent, for sale)",
        "Number of living rooms",
        "Number of bedrooms",
        "Number of bathrooms",
        "Surface area of the property (may include units)",
        "URL link to the listing"
    ],
    "Data Type": data_frame.dtypes.values,
    "Number of Missing Values": data_frame.isna().sum().values,
     "Number of Unique Values": data_frame.nunique().values
})

# Calculate preliminary statistics for numerical columns
numeric_data = data_frame.select_dtypes(include='number')
preliminary_stats = numeric_data.describe().T  # transpose for readability

# Calculate additional metrics: median
preliminary_stats['median'] = numeric_data.median()


In [17]:
#data catalogue 
data_catalog


Column                                       Description Data Type  \
0          titre                     Title of the property listing    object   
1           prix      Price of the property (may include currency)   float64   
2    emplacement                          Location of the property    object   
3  type_de_vente    Type of transaction (e.g., for rent, for sale)    object   
4          salon                            Number of living rooms   float64   
5        chambre                                Number of bedrooms   float64   
6       toilette                               Number of bathrooms   float64   
7        surface  Surface area of the property (may include units)   float64   
8           lien                           URL link to the listing    object   

   Number of Missing Values  Number of Unique Values  
0                         0                      402  
1                        40                      168  
2                         0                       32  
3                         0                        2  
4                       159                        5  
5                       297                        5  
6                       300                        5  
7                       154                       97  
8                         0                      414

In [19]:
preliminary_stats

count           mean            std    min    25%     50%       75%  \
prix      391.0  419259.439898  764585.663034  199.0  500.0  6800.0  680000.0   
salon     272.0       1.261029       0.590412    0.0    1.0     1.0       1.0   
chambre   134.0       2.410448       0.695989    1.0    2.0     2.0       3.0   
toilette  131.0       2.343511       7.127165    0.0    1.0     2.0       2.0   
surface   277.0     101.916968      47.957601    1.0   75.0    94.0     120.0   

                max  median  
prix      5500000.0  6800.0  
salon           4.0     1.0  
chambre         5.0     2.0  
toilette       83.0     2.0  
surface       504.0    94.0

# Analyse des résultats statistiques

Voici l'interprétation des résultats pour les différentes caractéristiques des biens immobiliers. Nous avons observé la moyenne, la médiane, l'écart-type et les autres statistiques descriptives.

| **Caractéristique** | **Compte (count)** | **Moyenne (mean)**  | **Écart-type (std)** | **Minimum (min)** | **25%** | **Médiane (50%)** | **75%** | **Maximum (max)** | **Interprétation** |
|---------------------|--------------------|---------------------|----------------------|-------------------|---------|-------------------|---------|-------------------|--------------------|
| **Prix**            | 391                | 419 259,44          | 764 585,66           | 199               | 500     | 6 800             | 680 000 | 5 500 000         | **Asymétrie à droite** : La moyenne est beaucoup plus élevée que la médiane, ce qui indique une longue queue vers les prix plus élevés. La majorité des prix sont proches du minimum ou autour de 6 800, mais il existe quelques biens très chers qui augmentent la moyenne. |
| **Salon**           | 272                | 1,26                | 0,59                 | 0                 | 1       | 1                 | 1       | 4                 | **Symétrie** : La moyenne et la médiane sont proches l'une de l'autre, et la majorité des biens ont un salon d'une taille standard (1 ou 0). |
| **Chambre**         | 134                | 2,41                | 0,70                 | 1                 | 2       | 2                 | 3       | 5                 | **Symétrie** : La distribution est plutôt symétrique, avec une moyenne proche de la médiane. La plupart des biens ont entre 2 et 3 chambres. |
| **Toilette**        | 131                | 2,34                | 7,13                 | 0                 | 1       | 2                 | 2       | 83                | **Asymétrie à droite** : Bien que la médiane soit 2 toilettes, l'écart-type élevé (7,13) et la présence de valeurs extrêmes (jusqu'à 83 toilettes) suggèrent une asymétrie avec quelques biens ayant beaucoup plus de toilettes que la majorité. |
| **Surface**         | 277                | 101,92              | 47,96                | 1                 | 75      | 94                | 120     | 504               | **Asymétrie à droite légère** : La moyenne est légèrement supérieure à la médiane, indiquant une légère asymétrie à droite. La majorité des surfaces se situent autour de 75 à 120 m², avec quelques propriétés très grandes. |

### Conclusion :
- **Prix** et **toilette** montrent une **forte asymétrie à droite**, avec des valeurs extrêmes qui influencent la moyenne.
- **Salon** et **chambre** montrent une **distribution symétrique**, ce qui signifie que la majorité des valeurs se concentrent autour de la médiane.
- **Surface** présente une légère **asymétrie à droite**, avec une concentration des surfaces dans une plage moyenne, mais avec quelques biens ayant des surfaces très grandes.

Ces informations nous aident à comprendre la répartition des différentes caractéristiques des propriétés et à identifier des tendances ou des anomalies dans les données.


In [23]:
# # Calcul des quartiles et de l'IQR pour les colonnes 'salon' et 'toilette'
# Q1_salon = data_frame['salon'].quantile(0.25)
# Q3_salon = data_frame['salon'].quantile(0.75)
# IQR_salon = Q3_salon - Q1_salon

# Q1_toilette = data_frame['toilette'].quantile(0.25)
# Q3_toilette = data_frame['toilette'].quantile(0.75)
# IQR_toilette = Q3_toilette - Q1_toilette

# # Détection des valeurs aberrantes pour 'salon' et 'toilette'
# outliers_salon = data_frame[(data_frame['salon'] < (Q1_salon - 1.5 * IQR_salon)) | (data_frame['salon'] > (Q3_salon + 1.5 * IQR_salon))]
# outliers_toilette = data_frame[(data_frame['toilette'] < (Q1_toilette - 1.5 * IQR_toilette)) | (data_frame['toilette'] > (Q3_toilette + 1.5 * IQR_toilette))]

# # Affichage des résultats
# print("Valeurs aberrantes pour 'salon' :\n", outliers_salon[['salon']])
# print("\nValeurs aberrantes pour 'toilette' :\n", outliers_toilette[['toilette']])


In [25]:

# Calcul des statistiques descriptives groupées par type de vente
grouped_statistics = data_frame.groupby('type_de_vente')[['prix', 'salon', 'chambre', 'toilette', 'surface']].describe()

# grouped_statistics['median'] = data_frame.groupby('type_de_vente')[['prix', 'salon', 'chambre', 'toilette', 'surface']].median()
# Réorganiser les colonnes pour une meilleure lisibilité
grouped_statistics = grouped_statistics.stack(level=0).rename_axis(['type_vente', 'statistiques']).reset_index()
grouped_statistics['median'] = grouped_statistics['50%']
# Afficher les résultats



C:\Users\Youcode\AppData\Local\Temp\ipykernel_15332\778125354.py:6: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  grouped_statistics = grouped_statistics.stack(level=0).rename_axis(['type_vente', 'statistiques']).reset_index()


In [27]:
grouped_statistics

type_vente statistiques  count          mean            std  \
0                 A louer      chambre    1.0  5.000000e+00            NaN   
1                 A louer         prix  258.0  2.977807e+04  348665.930018   
2                 A louer        salon  132.0  1.234848e+00       0.564332   
3                 A louer      surface  130.0  1.064462e+02      58.812679   
4                 A louer     toilette    1.0  3.000000e+00            NaN   
5  Appartements, à vendre      chambre  133.0  2.390977e+00       0.660975   
6  Appartements, à vendre         prix  133.0  1.174795e+06  786675.282373   
7  Appartements, à vendre        salon  140.0  1.285714e+00       0.614989   
8  Appartements, à vendre      surface  147.0  9.791156e+01      35.457272   
9  Appartements, à vendre     toilette  130.0  2.338462e+00       7.154501   

      min       25%       50%        75%        max    median  
0     5.0       5.0       5.0        5.0        5.0       5.0  
1   199.0     350.0    1050.0     6500.0  5500000.0    1050.0  
2     0.0       1.0       1.0        1.0        4.0       1.0  
3     1.0      70.0      95.0      120.0      504.0      95.0  
4     3.0       3.0       3.0        3.0        3.0       3.0  
5     1.0       2.0       2.0        3.0        4.0       2.0  
6  7500.0  660000.0  960000.0  1500000.0  4500000.0  960000.0  
7     0.0       1.0       1.0        1.0        4.0       1.0  
8    36.0      77.0      91.0      119.0      281.0      91.0  
9     0.0       1.0       2.0        2.0       83.0       2.0

# Analyse de la symétrie des données

L'objectif est de vérifier si les données sont symétriques ou asymétriques pour chaque type de vente. Nous avons comparé la moyenne et la médiane pour en déduire l'asymétrie. Voici les résultats :

| **Type de vente**           | **Statistique** | **Moyenne**   | **Médiane**   | **Interprétation**                                       |
|-----------------------------|-----------------|---------------|---------------|----------------------------------------------------------|
| **A louer**                 | Prix            | 29 778,07     | 1 050         | **Asymétrie à droite** : Moyenne > Médiane, fortement biaisé vers les valeurs élevées. |
| **A louer**                 | Surface         | 106,45        | 95            | **Légère asymétrie à droite** : Moyenne légèrement supérieure à la médiane. |
| **Appartements, à vendre**  | Prix            | 1 174 795     | 960 000       | **Asymétrie à droite** : Moyenne > Médiane, fortement biaisé vers les valeurs élevées. |
| **Appartements, à vendre**  | Surface         | 97,91         | 91            | **Légère asymétrie à droite** : Moyenne légèrement supérieure à la médiane. |

### Conclusion :
- **Prix** dans les deux types de vente montrent une **forte asymétrie à droite**, suggérant que la majorité des valeurs sont concentrées autour des prix plus bas, avec une longue queue à droite des prix plus élevés.
- **Surface** montre une **asymétrie légère à droite**, ce qui peut indiquer une légère tendance à avoir des surfaces plus grandes dans certaines propriétés.

Pour une analyse plus fine, il serait possible de calculer la **skewness** (asymétrie) pour chaque colonne afin de confirmer ces tendances.


In [30]:
a_vendre = data_frame[data_frame['type_de_vente'] == 'Appartements, à vendre']

# Créer un DataFrame pour les biens à louer
a_louer = data_frame[data_frame['type_de_vente'] == 'A louer']

In [32]:
a_louer.info()

<class 'pandas.core.frame.DataFrame'>
Index: 284 entries, 0 to 430
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   titre          284 non-null    object 
 1   prix           258 non-null    float64
 2   emplacement    284 non-null    object 
 3   type_de_vente  284 non-null    object 
 4   salon          132 non-null    float64
 5   chambre        1 non-null      float64
 6   toilette       1 non-null      float64
 7   surface        130 non-null    float64
 8   lien           284 non-null    object 
dtypes: float64(5), object(4)
memory usage: 22.2+ KB


In [34]:

# Calculer l'asymétrie pour les colonnes numériques de chaque type de vente
numerical_columns = ['prix', 'salon', 'chambre', 'toilette', 'surface']

# Calculer la skewness pour 'Appartements, à vendre'
skewness_a_vendre = a_vendre[numerical_columns].apply(skew, nan_policy='omit')
print("Skewness pour les Appartements à vendre:")
print(skewness_a_vendre)

# Calculer la skewness pour 'A louer'
skewness_a_louer = a_louer[numerical_columns].apply(skew, nan_policy='omit')
print("\nSkewness pour les Appartements à louer:")
print(skewness_a_louer)

Skewness pour les Appartements à vendre:
prix         1.439070
salon        2.349000
chambre     -0.147229
toilette    11.149373
surface      1.252748
dtype: float64

Skewness pour les Appartements à louer:
prix        15.167855
salon        1.786307
chambre           NaN
toilette          NaN
surface      3.261515
dtype: float64


| Type de Vente            | Variable  | Skewness | Interprétation                                                                                             |
|--------------------------|-----------|----------|------------------------------------------------------------------------------------------------------------|
| Appartements à vendre    | Prix      | 1.44     | Distribution asymétrique à droite, avec des valeurs de prix plus élevées que la moyenne.                    |
|                          | Salon     | 2.35     | Fortement asymétrique à droite, la plupart des biens ont peu de salons, avec quelques-uns en ayant plusieurs. |
|                          | Chambre   | -0.15    | Faiblement asymétrique à gauche, indiquant une distribution relativement équilibrée des chambres.           |
|                          | Toilette  | 11.15    | Extrêmement asymétrique à droite, très peu de biens ont plusieurs toilettes.                                |
|                          | Surface   | 1.25     | Distribution asymétrique à droite, certaines surfaces sont bien plus grandes que la moyenne.                |
| Appartements à louer     | Prix      | 15.17    | Très fortement asymétrique à droite, avec quelques prix très élevés par rapport à la majorité.              |
|                          | Salon     | 1.79     | Asymétrique à droite, la plupart des biens ont peu de salons, certains en ont plusieurs.                    |
|                          | Chambre   | NaN      | Valeur non calculable (données manquantes ou insuffisantes).                                               |
|                          | Toilette  | NaN      | Valeur non calculable (données manquantes ou insuffisantes).                                               |
|                          | Surface   | 3.26     | Fortement asymétrique à droite, indiquant quelques biens avec une grande surface.                           |


In [37]:
a_louer.columns = a_louer.columns.str.capitalize()
a_vendre.columns = a_vendre.columns.str.capitalize()

In [39]:
from sklearn.impute import SimpleImputer
# Utilisation de .loc pour éviter le SettingWithCopyWarning

a_louer.loc[:, 'Prix'] = imputer_prix.fit_transform(a_louer[['Prix']])
a_vendre.loc[:, 'Prix'] = imputer_prix.fit_transform(a_vendre[['Prix']])

a_louer.loc[:, 'Salon'] = imputer_salon.fit_transform(a_louer[['Salon']])
a_vendre.loc[:, 'Salon'] = imputer_salon.fit_transform(a_vendre[['Salon']])

a_louer.loc[:, 'Chambre'] = imputer_chambre.fit_transform(a_louer[['Chambre']])
a_vendre.loc[:, 'Chambre'] = imputer_chambre.fit_transform(a_vendre[['Chambre']])

a_louer.loc[:, 'Toilette'] = imputer_toilette.fit_transform(a_louer[['Toilette']])
a_vendre.loc[:, 'Toilette'] = imputer_toilette.fit_transform(a_vendre[['Toilette']])

a_louer.loc[:, 'Surface'] = imputer_surface.fit_transform(a_louer[['Surface']])
a_vendre.loc[:, 'Surface'] = imputer_surface.fit_transform(a_vendre[['Surface']])


NameError: name 'imputer_prix' is not defined

In [146]:
a_vendre.describe()

Prix       Salon     Chambre    Toilette     Surface
count  1.330000e+02  140.000000  133.000000  130.000000  147.000000
mean   1.174795e+06    1.285714    2.390977    2.338462   97.911565
std    7.866753e+05    0.614989    0.660975    7.154501   35.457272
min    7.500000e+03    0.000000    1.000000    0.000000   36.000000
25%    6.600000e+05    1.000000    2.000000    1.000000   77.000000
50%    9.600000e+05    1.000000    2.000000    2.000000   91.000000
75%    1.500000e+06    1.000000    3.000000    2.000000  119.000000
max    4.500000e+06    4.000000    4.000000   83.000000  281.000000

In [61]:
a_louer.describe()

Prix       Salon  Chambre  Toilette     Surface
count  2.580000e+02  132.000000      1.0       1.0  130.000000
mean   2.977807e+04    1.234848      5.0       3.0  106.446154
std    3.486659e+05    0.564332      NaN       NaN   58.812679
min    1.990000e+02    0.000000      5.0       3.0    1.000000
25%    3.500000e+02    1.000000      5.0       3.0   70.000000
50%    1.050000e+03    1.000000      5.0       3.0   95.000000
75%    6.500000e+03    1.000000      5.0       3.0  120.000000
max    5.500000e+06    4.000000      5.0       3.0  504.000000

In [ ]:
# 1. Identifier le pourcentage de valeurs manquantes
missing_percentage = data_frame.isnull().mean() * 100
print("Pourcentage de valeurs manquantes par colonne :")
print(missing_percentage[missing_percentage > 0])  # Afficher uniquement les colonnes avec des valeurs manquantes

# 2. Choisir la méthode d'imputation
for column in data_frame.columns:
    if data_frame[column].isnull().any():  # Vérifier si la colonne a des valeurs manquantes
        # Calculer les statistiques descriptives pour décider de la méthode d'imputation
        mean = data_frame[column].mean()
        median = data_frame[column].median()
        std_dev = data_frame[column].std()

        # Choisir la méthode d'imputation
        if abs(mean - median) > std_dev:  # Distribution asymétrique
            imputation_value = median
            method = "médiane"
        else:  # Distribution symétrique
            imputation_value = mean
            method = "moyenne"

        # Imputer les valeurs manquantes
        data_frame[column].fillna(imputation_value)
        print(f"Imputation de la colonne '{column}' avec la {method} : {imputation_value}")


In [41]:
#tache 3

import pandas as pd
import numpy as np
from scipy.stats import zscore
from sklearn.preprocessing import StandardScaler, MinMaxScaler

# Détection des valeurs aberrantes avec le score Z
def detect_outliers_z_score(data, threshold=3):
    z_scores = zscore(data.select_dtypes(include=[np.number]))
    abs_z_scores = np.abs(z_scores)
    outliers = (abs_z_scores > threshold)
    print("Valeurs aberrantes détectées avec le score Z :")
    return data[outliers.any(axis=1)]

# Détection des valeurs aberrantes avec la méthode IQR
def detect_outliers_iqr(data, column):
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
    print(f"\nValeurs aberrantes dans la colonne '{column}' avec la méthode IQR :")
    return outliers

# 1. Exemple d'utilisation du score Z pour détecter les valeurs aberrantes
outliers_z_score = detect_outliers_z_score(data_frame)
print(outliers_z_score)

# 2. Exemple d'utilisation de la méthode IQR pour détecter les valeurs aberrantes dans chaque colonne
for column in data_frame.select_dtypes(include=[np.number]).columns:
    outliers_iqr = detect_outliers_iqr(data_frame, column)
    print(outliers_iqr)

# Standardisation des données (score Z)
scaler = StandardScaler()
data_standardized = pd.DataFrame(scaler.fit_transform(data_frame.select_dtypes(include=[np.number])),
                                 columns=data_frame.select_dtypes(include=[np.number]).columns)
print("\nDonnées standardisées :")
data_standardized


Valeurs aberrantes détectées avec le score Z :
Empty DataFrame
Columns: [titre, prix, emplacement, type_de_vente, salon, chambre, toilette, surface, lien]
Index: []

Valeurs aberrantes dans la colonne 'prix' avec la méthode IQR :
                                                 titre       prix  \
5              Appartement à vendre 127m² à Casablanca  2200000.0   
43     Un appartement de 155m² à Iberia avec 2 façades  2650000.0   
44                 Appartement à vendre 98 m² à Agadir  2400000.0   
59                 Appartement à vendre Hay Riad Rabat  3550000.0   
100                             Joli appart حي الإداري  2300000.0   
136  Appartement de 151m² en centre de Tanger à vendre  1850000.0   
160                     Vente Appartement Wifak Témara  1800000.0   
226          Appartement à vendre 80 m² à rue de rabat  4500000.0   
229    CMN-HA-1350 - Appartement à vendre à Beauséjour  1800000.0   
287          COUP DE CŒUR À AGDAL appart neuf de 96 m²  1750000.0   
289        

prix     salon   chambre  toilette   surface
0   -0.541848 -0.442929       NaN       NaN -0.040044
1    1.284352 -0.442929  0.850249 -0.048383  0.628419
2   -0.540539 -0.442929       NaN       NaN  0.294187
3   -0.534646 -0.442929       NaN       NaN  1.213325
4   -0.538575 -0.442929       NaN       NaN  0.210629
..        ...       ...       ...       ...       ...
426 -0.548265       NaN       NaN       NaN       NaN
427  0.315267  1.253926 -0.591946 -0.048383 -0.228050
428       NaN       NaN       NaN       NaN       NaN
429  1.218873 -0.442929 -0.591946 -0.048383 -0.436945
430       NaN       NaN       NaN       NaN       NaN

[431 rows x 5 columns]

In [ ]:

# Normalisation des données (entre 0 et 1)
min_max_scaler = MinMaxScaler()
data_normalized = pd.DataFrame(min_max_scaler.fit_transform(data_frame.select_dtypes(include=[np.number])),
                               columns=data_frame.select_dtypes(include=[np.number]).columns)
print("\nDonnées normalisées :")
data_normalized
